src/spreadinterp.cpp

```diff
@@ -777,24 +779,99 @@ Two upsampfacs implemented. Params must match ref formula. Barnett 4/24/18 */
+             i += simd_size, offset -= simd_size) {
+          k_odd  = if_odd_degree ? simd_type::load_aligned(padded_coeffs[0].data() + i)
+                                 : zerov;
+          k_even = simd_type::load_aligned(padded_coeffs[if_odd_degree].data() + i);
+          for (uint8_t j = 1 + if_odd_degree; j < nc; j += 2) {
+            const auto cji_odd = simd_type::load_aligned(padded_coeffs[j].data() + i);
+            k_odd              = xsimd::fma(k_odd, z2v, cji_odd);
+            const auto cji_even =
+                simd_type::load_aligned(padded_coeffs[j + 1].data() + i);
+            k_even = xsimd::fma(k_even, z2v, cji_even);
+          }
+          // left part
+          xsimd::fma(k_odd, zv, k_even).store_aligned(ker + i);
+          // right part symmetric to the left part
+          if (offset >= (w + 1) / 2) {
+            // to use aligned store, we need shuffle the previous k_sym and current k_sym
+            k_prev = k_sym;
+            k_sym  = xsimd::fma(k_odd, -zv, k_even);
+            xsimd::shuffle(k_sym, k_prev, shuffle_batch).store_aligned(ker + offset);
+          }
+        }
+      } else {
+        // xsimd constants for reverse
+        static constexpr auto reverse_batch =
+            xsimd::make_batch_constant<xsimd::as_unsigned_integer_t<FLT>, arch_t,
+                                       reverse_index<simd_size>>();
+
+        // process simd vecs
+        for (uint8_t i = 0, offset = w - simd_size; i < w / 2;
+             i += simd_size, offset -= simd_size) {
+          auto k_odd  = if_odd_degree
+                            ? simd_type::load_aligned(padded_coeffs[0].data() + i)
+                            : zerov;
+          auto k_even = simd_type::load_aligned(padded_coeffs[if_odd_degree].data() + i);
+          for (uint8_t j = 1 + if_odd_degree; j < nc; j += 2) {
+            const auto cji_odd = simd_type::load_aligned(padded_coeffs[j].data() + i);
+            k_odd              = xsimd::fma(k_odd, z2v, cji_odd);
+            const auto cji_even =
+                simd_type::load_aligned(padded_coeffs[j + 1].data() + i);
+            k_even = xsimd::fma(k_even, z2v, cji_even);
+          }
+          // left part
+          xsimd::fma(k_odd, zv, k_even).store_aligned(ker + i);
+          // right part symmetric to the left part
+          if (offset >= w / 2) {
+            // reverse the order for symmetric part
+            xsimd::swizzle(xsimd::fma(k_odd, -zv, k_even), reverse_batch)
+                .store_aligned(ker + offset);
+          }
+        }
+      }
+    } else {
+      const simd_type zv(z);
+
+      for (uint8_t i = 0; i < w; i += simd_size) {
+        auto k = simd_type::load_aligned(padded_coeffs[0].data() + i);
+        for (uint8_t j = 1; j < nc; ++j) {
+          const auto cji = simd_type::load_aligned(padded_coeffs[j].data() + i);
+          k              = xsimd::fma(k, zv, cji);
+        }
+        k.store_aligned(ker + i);
```

Comment:
there is quite a lot of repetition I think you could do something like w/2 + (tail>0) to merge most of the code.

Reply:
Could you give more of a hint what you're suggesting here? Or maybe Libin sees it...